### PR TITLE
result stats: Enable persisting ingestion/query benchmark results in a common format across targets

### PR DESCRIPTION
**Pull request by filipecosta90**
This PR kicks off the discussion about persisting ingestion/querying benchmark result stats in a common format.
We've added the samples to full_cycle_minitest.

To do so, we've added `--results-file` to both loader and query runner tools. If no `--result-file` is specified then no summary result JSON is stored. 
Bellow you can find the sample output jsons for both loading and querying after running `PASSWORD="password" ./scripts/full_cycle_minitest/full_cycle_minitest_timescaledb.sh` with TimeScaleDB locally available on default ports.

**Ingestion result sample json output**

```json
{
 "ResultFormatVersion": "0.1",
 "RunnerConfig": {
  "db-name": "benchmark",
  "batch-sze": 10000,
  "workers": 1,
  "limit": 0,
  "do-load": true,
  "do-create-db": true,
  "do-abort-on-exist": false,
  "reporting-period": 10000000000,
  "hash-workers": false,
  "no-flow-control": false,
  "channel-capacity": 0,
  "insert-intervals": "",
  "results-file": "timescaledb_load_results.json",
  "file": "/tmp/bulk_data/timescaledb_data",
  "seed": 0
 },
 "StartTime": 1620126088,
 "EndTime": 1620126090,
 "DurationMillis": 1312,
 "Totals": {
  "metricRate": 658166.850104776,
  "rowRate": 65816.6850104776
 }
}
```

**query result sample json output**
```json
{
 "ResultFormatVersion": "0.1",
 "RunnerConfig": {
  "DBName": "benchmark",
  "Limit": 1000,
  "LimitRPS": 0,
  "MemProfile": "",
  "HDRLatenciesFile": "0rps_timescaledb_query_high-cpu-1.hdr",
  "Workers": 1,
  "PrintResponses": false,
  "Debug": 0,
  "FileName": "/tmp/bulk_data/timescaledb_query_high-cpu-1",
  "BurnIn": 0,
  "PrintInterval": 100,
  "PrewarmQueries": false,
  "ResultsFile": "timescaledb_query_high-cpu-1_results.json"
 },
 "StartTime": 1620126095,
 "EndTime": 1620126100,
 "DurationMillis": 4361,
 "Totals": {
  "burnIn": 0,
  "limit": 1000,
  "overallQuantiles": {
   "TimescaleDB CPU over threshold, 1 host(s)": {
    "q0": 0,
    "q100": 16.571,
    "q50": 4.039,
    "q95": 7.625,
    "q99": 9.272,
    "q999": 12.065
   },
   "allqueries": {
    "q0": 0,
    "q100": 16.571,
    "q50": 4.039,
    "q95": 7.625,
    "q99": 9.272,
    "q999": 12.065
   }
  },
  "overallQueryRates": {
   "TimescaleDB CPU over threshold, 1 host(s)": 229.3165254507947,
   "allqueries": 229.3165254507947
  },
  "prewarmQueries": false
 }
}
```

### CODE REVIEW

**Review by jonatas — APPROVED** [review]
Tested manually! The JSON output looks great! thanks a lot for this contribution!

I hope @RobAtticus can merge it later today :)

File: load/loader.go [hunk 1]
```diff
@@ -29,21 +31,22 @@ var (
...
-	DoLoad          bool          `yaml:"do-load" mapstructure:"do-load"`
-	DoCreateDB      bool          `yaml:"do-create-db" mapstructure:"do-create-db"`
-	DoAbortOnExist  bool          `yaml:"do-abort-on-exist" mapstructure:"do-abort-on-exist"`
-	ReportingPeriod time.Duration `yaml:"reporting-period" mapstructure:"reporting-period"`
-	HashWorkers     bool          `yaml:"hash-workers" mapstructure:"hash-workers"`
-	NoFlowControl   bool          `yaml:"no-flow-control" mapstructure:"no-flow-control"`
-	ChannelCapacity uint          `yaml:"channel-capacity" mapstructure:"channel-capacity"`
-	InsertIntervals string        `yaml:"insert-intervals" mapstructure:"insert-intervals"`
+	DBName          string        `yaml:"db-name" mapstructure:"db-name" json:"db-name"`
+	BatchSize       uint          `yaml:"batch-size" mapstructure:"batch-size" json:"batch-sze"`
```
Comment (antekresic) [inline load/loader.go:35]:
```suggestion
	BatchSize       uint          `yaml:"batch-size" mapstructure:"batch-size" json:"batch-size"`
```

File: load/loader.go [hunk 2]
```diff
@@ -138,7 +142,41 @@ func (l *CommonBenchmarkRunner) postRun(wg *sync.WaitGroup, start *time.Time) {
...
+		l.saveTestResult(took, *start, end)
+	}
+}
+
+func (l *CommonBenchmarkRunner) saveTestResult(took time.Duration, start time.Time, end time.Time) {
+	totals := make(map[string]interface{})
+	metricRate := float64(l.metricCnt) / took.Seconds()
+	totals["metricRate"] = metricRate
+	if l.rowCnt > 0 {
+		rowRate := float64(l.rowCnt) / float64(took.Seconds())
```
Comment (antekresic) [inline load/loader.go:157]: `took.Seconds()` returns a `float64`, unnecessary conversion.

File: pkg/query/stat_processor.go [hunk 3]
```diff
@@ -188,6 +194,70 @@ func (sp *defaultStatProcessor) process(workers uint) {
...
+	configs := map[string]interface{}{}
+	_, all := generateQuantileMap(histogram)
+	configs[label] = all
+	configs["EncodedHistogram"] = nil
+	encodedHist, err := histogram.Encode(hdrhistogram.V2CompressedEncodingCookieBase)
+	if err == nil {
+		configs["EncodedHistogram"] = encodedHist
+	}
+	return configs
+}
```
Comment (antekresic) [inline pkg/query/stat_processor.go:228]: Seems like an unused function, what is reason for adding it?

File: pkg/query/stat_processor.go [hunk 4]
```diff
@@ -174,7 +180,7 @@ func (sp *defaultStatProcessor) process(workers uint) {
 		_, _ = fmt.Printf("Saving High Dynamic Range (HDR) Histogram of Response Latencies to %s\n", sp.args.hdrLatenciesFile)
 		var b bytes.Buffer
 		bw := bufio.NewWriter(&b)
-		_, err = statMapping[allQueriesLabel].latencyHDRHistogram.PercentilesPrint(bw, 10, 1000.0)
+		_, err = sp.statMapping[labelAllQueries].latencyHDRHistogram.PercentilesPrint(bw, 10, 1000.0)
```
Comment (antekresic) [inline pkg/query/stat_processor.go:183]: Whats the reason for switch from `allQueriesLabel` const? It seems to be used in the code of the function prior to this line.

File: load/loader.go [hunk 5]
```diff
@@ -138,7 +142,41 @@ func (l *CommonBenchmarkRunner) postRun(wg *sync.WaitGroup, start *time.Time) {
 	// Wait for all workers to finish
 	wg.Wait()
 	end := time.Now()
-	l.summary(end.Sub(*start))
+	took := end.Sub(*start)
+	l.summary(took)
+	if len(l.BenchmarkRunnerConfig.ResultsFile) > 0 {
```
Comment (antekresic) [inline load/loader.go:147]:
```suggestion
	if l.BenchmarkRunnerConfig.ResultsFile != "" {
```
Check against default value.

File: load/loader.go [hunk 6]
```diff
@@ -138,7 +142,41 @@ func (l *CommonBenchmarkRunner) postRun(wg *sync.WaitGroup, start *time.Time) {
...
+}
+
+func (l *CommonBenchmarkRunner) saveTestResult(took time.Duration, start time.Time, end time.Time) {
+	totals := make(map[string]interface{})
+	metricRate := float64(l.metricCnt) / took.Seconds()
+	totals["metricRate"] = metricRate
+	if l.rowCnt > 0 {
+		rowRate := float64(l.rowCnt) / float64(took.Seconds())
+		totals["rowRate"] = rowRate
+	}
```
Comment (antekresic) [inline load/loader.go:159]: This part of the function does result transformation which should not be a part of the function whose purpose is to save the test result.

I suggest extracting it outside of this function.

**Review by antekresic — CHANGES_REQUESTED** [review]

File: pkg/query/stat_processor.go (hunk 3 above)
Comment (filipecosta90) [reply pkg/query/stat_processor.go]: @antekresic you're totally right. I've copied it from other tools we use and forgot to delete this unsed function.  addressed in latest commit

File: pkg/query/stat_processor.go (hunk 4 above)
Comment (filipecosta90) [reply pkg/query/stat_processor.go]: typo. addressed in latest commit

File: load/loader.go (hunk 6 above)
Comment (filipecosta90) [reply load/loader.go]: addressed in latest commit

**Comment by filipecosta90** [comment]
@jonatas @antekresic I believe all comments are addressed on the latest commit. Can you guys check it?

**Review by antekresic — APPROVED** [review]

**Pull request merged**